### PR TITLE
Explore: handle URLs parse errors

### LINKED ISSUE #74687: Explore: Create more informative error page when a link doesn't parse
Internal conversation about this is here: https://raintank-corp.slack.com/archives/C036J5B39/p1694448696250129

In the event an explore URL cannot be parsed, we should fail more gracefully - showing the URL as much as we can parse out of it. Often, recreating a query is relatively trivial, but that cant be done if the page crashes and a refresh brings us to the default datasource.

### PR DESCRIPTION
**Pull request by Elfo404**
**What is this feature?**

Handles URL parsing errors in Explore and shows a notification instead of crashing the page when a user opens explore with a a malformed URL.

@diegoadams do we want to go with the notification (see screenshot) or some sort of dismissable box somewhere in the page? (ignore the wording in the screenshot, we have def better options)
image

**Which issue(s) does this PR fix?**:

Fixes #74687 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by diegoadams** [comment]
Hey @Elfo404, I think the toast notification works well for this scenario. Since there's no immediate action required from the user and the purpose is just to inform them about the error, it conveys the message without disrupting their workflow.

**Comment by GrafanaWriter** [comment]
Removing the `type/docs` label as this label triggers a GH action that adds this to the Docs org-wide project and it's not clear that there's a tech docs impact from this change. Please let me know if that's incorrect.

**Review by harisrozajac — APPROVED** [review]
This works great! 🚀  I'm curious what is the difference between "Could not parse URL" that is implemented here and the "Query processing error - Malformed URI" one. Is it that the second one shows when queries param is malformed?

I've tested various versions of messing up the URL and sometimes I get the first message and sometimes the second one.

**Comment by Elfo404** [comment]
@GrafanaWriter there is a change in `docs/sources/explore/_index.md`

**Comment by Elfo404** [comment]
@harisrozajac i'm not entirely sure where that specific error originates from, but that happens when running a query returns an error. Most likely because you messed up with something in the URL that makes the URL still structurally valid, but the underlying datasource errors out. do you have an example?

File: docs/sources/explore/_index.md [hunk 1]
```diff
@@ -83,6 +83,10 @@ You can then click on any panel icon in the content outline to navigate to that
 
 When using Explore, the URL in the browser address bar updates as you make changes to the queries. You can share or bookmark this URL.
 
+{{% admonition type="note" %}}
+Explore may generate relatively long URLs, some tools, like messaging or videoconferencing apps, may truncate messages to a fixed length. In such cases Explore will display a warning message and load a default state. If you encounter issues when sharing Explore links in such apps, you can generate shortened links. See [Share shortened link]("#share-shortened-link"}}) for more information.
```
Comment (ifrost) [inline docs/sources/explore/_index.md:87]: This link doesn't work. Generates a URL to http://localhost:3002/docs/grafana/latest/explore/%22#share-shortened-link%22%7d%7d

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 2]
```diff
@@ -248,8 +255,9 @@ export function useStateSync(params: ExploreQueryParams) {
               acc[key] = undefined;
               return acc;
             }, {}),
-            panes: JSON.stringify(newParams.panes),
+            // we set the schemaVersion as the first parameter so that when URLs are truncated the schemaVersion is more likely to be present.
```
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync/index.ts:258]: Doesn't seem to work every time 🤔 


https://github.com/grafana/grafana/assets/745532/178fce75-4178-4458-8304-7da34b689cc4

**Comment by harisrozajac** [comment]
> Most likely because you messed up with something in the URL that makes the URL still structurally valid, but the underlying datasource errors out. do you have an example?

@Elfo404 sure, here are the examples:

Full, working URL: https://ephemeral1511182177784elfo404.grafana-dev.net/explore?schemaVersion=1&panes=%7B%228kX%22:%7B%22datasource%22:%22grafanacloud-usage%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22grafanacloud_grafana_instance_alerting_alertmanager_alerts%7Borg_id%3D%5C%22916%5C%22%7D%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22grafanacloud-usage%22%7D,%22editorMode%22:%22builder%22,%22legendFormat%22:%22__auto%22,%22useBackend%22:false,%22disableTextWrap%22:false,%22fullMetaSearch%22:false,%22includeNullMetadata%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1

Errors that happen as I mess up the URL:
1. Query Processing Error - URI malformed (without queries param): https://ephemeral1511182177784elfo404.grafana-dev.net/explore?schemaVersion=1&panes=%7B%22XHj%22:%7B%22datasource%22:%22grafanacloud-usage%22,%22queriesadasdasdas%22:%5B%7B%22refId%22dasdasdasda%22A%22,%22expr%22:%22%22,%22edsadasdasdas%22:true,%22instant%22:true,%adasdasd2datasource%22sadasdada:%7B%22type%22:%22prometheus%dasdasdasda2,%22uid%22:%22grafanacloud-usage%22%7D%7D%5D,%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1dasdasdasd
2. Query Processing Error - URI malformed (with queries param): https://ephemeral1511182177784elfo404.grafana-dev.net/explore?schemaVersion=1&panes=%7B%228kX%22:%7B%22datasource%22:%22grafanacloud-usage%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22grafanacloud_grafana_instance_alerting_alertmanager_alerts%7%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1
3. Could not parse Explore  URL: [https://ephemeral1511182177784elfo404.grafana-dev.net/explore?schemaVersion=1&panes=%7B%228kX%22:%7B%22datasource%22:%22grafanacloud-usage%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22grafanacloud_grafana_instance_alerting_alertmanager_alerts%7Borg_id%3D%5C%22916%5C%22%7D%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22grafanacloud-usage%22%7D,%22editorMode%22:%22builder%22,%22legendFormat%22:%22__auto%22,%22useBackend%22:false,%22disableTextWrap%22:false,](https://ephemeral1511182177784elfo404.grafana-dev.net/explore?schemaVersion=1&panes=%7B%228kX%22:%7B%22datasource%22:%22grafanacloud-usage%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22grafanacloud_grafana_instance_alerting_alertmanager_alerts%7Borg_id%3D%5C%22916%5C%22%7D%22,%22range%22:true,%22instant%22:true,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22grafanacloud-usage%22%7D,%22editorMode%22:%22builder%22,%22legendFormat%22:%22__auto%22,%22useBackend%22:false,%22disableTextWrap%22:false,%22fullMetaSearch%22:false,%22includeNullMetadata%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-6h%22,%22to%22:%22now%22%7D%7D%7D&orgId=1)

File: public/app/features/explore/hooks/useStateSync/index.ts (hunk 2 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync/index.ts]: amazing catch! 
this happens when the query completes before Explore has done initializing (?)

i've pushed a tentative fix in https://github.com/grafana/grafana/pull/77784/commits/959a83a3c2c627e2306b3128524d93512f81eebf

File: docs/sources/explore/_index.md (hunk 1 above)
Comment (Elfo404) [reply docs/sources/explore/_index.md]: should be fixed now, mind double checking?

File: public/app/features/explore/hooks/useExplorePageTitle.ts [hunk 3]
```diff
@@ -19,8 +18,19 @@ export function useExplorePageTitle(params: ExploreQueryParams) {
       return;
     }
 
+    let panesObject: unknown;
```
Comment (ifrost) [inline public/app/features/explore/hooks/useExplorePageTitle.ts:21]: Could we modify `safeParseJson` instead? Or at least delete it? I think that was the only place it was used.

File: public/app/features/explore/hooks/useStateSync/index.ts [hunk 4]
```diff
@@ -76,8 +78,11 @@ export function useStateSync(params: ExploreQueryParams) {
           if (!isEqual(prevParams.current.panes, JSON.stringify(panesQueryParams))) {
             // If there's no previous state it means we are mounting explore for the first time,
             // in this case we want to replace the URL instead of pushing a new entry to the history.
+            // If theinitstate is 'pending' it means explore still hasn't finished initializing. in that case we skip
```
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync/index.ts:81]:
```suggestion
            // If the init state is 'pending' it means explore still hasn't finished initializing. in that case we skip
```

File: public/app/features/explore/hooks/useStateSync/migrators/types.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@ export interface MigrationHandler<From extends BaseExploreURL | never, To> {
   /**
    * The parse function is used to parse the URL parameters into the state object.
    */
-  parse: (params: UrlQueryMap) => To;
+  parse: (params: UrlQueryMap) => [To, boolean];
```
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync/migrators/types.ts:7]: nit: I think it'd be more readable if we returned an object not an array

File: public/app/features/explore/hooks/useStateSync/migrators/v0.ts [hunk 6]
```diff
@@ -12,13 +11,52 @@ export interface ExploreURLV0 extends BaseExploreURL {
...
+      left = FALLBACK_PANE_VALUE;
+    }
+
+    return [
+      {
+        schemaVersion: 0,
+        left,
+        ...(right && { right }),
+      },
+      leftError || !!rightError,
```
Comment (ifrost) [inline public/app/features/explore/hooks/useStateSync/migrators/v0.ts:58]: Why not:

```suggestion
      !!leftError || !!rightError,
```

**Review by ifrost — APPROVED** [review]
Looks good. Left a few minor suggestions.

File: public/app/features/explore/hooks/useExplorePageTitle.ts (hunk 3 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useExplorePageTitle.ts:21]: i didn't notice it wasn't used anymore, honestly i'd rather remove it as there's nothing "safe" about it imho

File: public/app/features/explore/hooks/useStateSync/migrators/types.ts (hunk 5 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync/migrators/types.ts]: agree!

File: public/app/features/explore/hooks/useStateSync/migrators/v0.ts (hunk 6 above)
Comment (Elfo404) [reply public/app/features/explore/hooks/useStateSync/migrators/v0.ts]: changed it for clarity but it's not really needed
`undefined || something` always evaluates to `something`, that's why i doubled banged only `rightError`

**Pull request merged**